cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -244,3 +275,108 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
 	}
 	return nil
 }
+
+// translateStubDomainOfKubeDNSToProxyCoreDNS translates StubDomain Data in kube-dns ConfigMap
+// in the form of Proxy for the CoreDNS Corefile.
+func translateStubDomainOfKubeDNSToProxyCoreDNS(dataField string, kubeDNSConfigMap *v1.ConfigMap) (string, error) {
+	if proxy, ok := kubeDNSConfigMap.Data[dataField]; ok {
+		stubDomainData := make(map[string][]string)
+		var proxyStanzaList string
+
+		err := json.Unmarshal([]byte(proxy), &stubDomainData)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse JSON from 'kube-dns ConfigMap: %v", err)
+		}
+		var proxyStanza []interface{}
+
+		for domain, proxyIP := range stubDomainData {
+			strings.Join(proxyIP, " ")
+			pStanza := map[string]interface{}{}
+			pStanza["keys"] = []string{domain + ":53"}
+			pStanza["body"] = [][]string{
+				{"errors"},
+				{"cache", "30"},
+				{"proxy", ".", strings.Join(proxyIP, " ")},
+			}
+
+			proxyStanza = append(proxyStanza, pStanza)
+		}
+		stanzasBytes, err := json.Marshal(proxyStanza)
+		if err != nil {
+			return "", err
+		}
+
+		outputJSON, err := caddyfile.FromJSON(stanzasBytes)
+		if err != nil {
+			return "", err
+		}
+		// This is required to format the Corefile, otherwise errors due to bad yaml format.
+		output := strings.NewReplacer("\n\t", "\n        ", "\"", "", "\n}", "\n    }", "\n\n", "\n    ").Replace(string(outputJSON))
```

Comment:
This seems too crude. Why not use BASE64 or double-encode as JSON?

Reply:
The reason I had to use the string replacer is because of the difference in formatting in JSON and YAML.
Since the manifest is parsed as a yaml, it errors out due to tabs instead of spaces and so on….

Reply:
I understand why you ended up having to do this, but this is a very crude method. Have you tried double encoding it?

Reply:
I am not sure I understand fully, but the double encoding would avoid the replacement of \" .. not really the formatting part - I mean we still need the replacement of \t, \n with space etc ...

Looks like there is no nice way to use the original codec (caddyfile). Because we then endup with incompatibility between JSON and YAML, and then we have anyway the formatting concern because that config can be edited directly as kubectl cmd. (we want it to be kind-of human readable format).

So what is the best to do here, without over complicating the code - that is, at the end, used only once. ?

Go back to initial code with simple string translation ?

@errordeveloper : what do you think ?

Reply:
Why not use [caddyfile](https://godoc.org/github.com/mholt/caddy/caddyfile)? It's the official package for parsing caddyfiles and it's self-contained (only stdlib code).

Reply:
Oh, it's already used above. Can't you use https://github.com/kubernetes/kubernetes/tree/master/vendor/github.com/ghodss/yaml to go from JSON to YAML?

Reply:
yaml.JSONToYAML

Reply:
I cannot because the output is in the form of caddyfile text, which will be parsed into the existing Corefile in the configmap.
(The variable name outputJSON is misleading, I'll fix that)